### PR TITLE
Import Journals

### PR DESCRIPTION
**Pull request by casaper**
# [Story in Trello](https://trello.com/c/V7AHWKS3/31-2-text-eintr%C3%A4ge-journale-importieren-korrespondenz)

### Done for approval?

* [x] Acceptance criteria are met
* [x] Code quality checks are green
* [ ] ~Readme updated if needed~
* [ ] Story under test
* [ ] ~Mobile works~
* [ ] ~Seeds created~
* [ ] ~Translated~
* [x] Code is reviewed

### Done after the merge?

* [ ] Deployed to staging after the merge
* [ ] Tested manually on staging

### CODE REVIEW

File: .overcommit.yml [hunk 1]
```diff
@@ -22,7 +22,7 @@ PreCommit:
     problem_on_unmodified_line: warn
 
   RailsSchemaUpToDate:
-    enabled: true
+    enabled: false
```
Comment (chrysalag) [inline .overcommit.yml]: Maybe that should go in a separate PR.

File: lib/access_import/access_import.rb [hunk 2]
```diff
@@ -75,6 +76,25 @@ def make_assignments
       assignment.creator_id = @import_user.id
       assignment.save!
     end
+    puts "Imported #{Assignment.count - records_before} new #{Assignment.class.name} "\
+      'from MS Access Database.'
```
Comment (toupeira) [inline lib/access_import/access_import.rb]: More readable:

```ruby
puts "Imported %d new %s from MS Access Database." %
  [Assignment.count - records_before, Assignment.name]
```

(you probably meant `Assignment.name`, `Assignment.class.name` is always "Class" ;-))

File: lib/access_import/journal_transform.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+require 'acc_utils'
+
+class JournalTransform
+  include AccUtils
+
+  def initialize(*accessors)
+    accessors.each do |accessor|
+      instance_variable_set("@#{accessor.class.name.underscore}", accessor)
+    end
```
Comment (toupeira) [inline lib/access_import/journal_transform.rb]: This doesn't look like it's actually used?

File: lib/access_import/journal_transform.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  def prepare_attributes(journal, person, assignment, user_id)
+    {
+      body: journal[:m_Text],
+      journalable_type: person.class.name,
+      journalable_id: person.id,
```
Comment (toupeira) [inline lib/access_import/journal_transform.rb]: You can probably just do `journalable: person`?

File: lib/access_import/journal_transform.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      assignment_id: assignment&.id,
```
Comment (toupeira) [inline lib/access_import/journal_transform.rb]: And `assignment: assignment`

File: lib/access_import/journal_transform.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+      created_at: journal[:d_ErfDatum],
+      updated_at: journal[:d_MutDatum],
```
Comment (toupeira) [inline lib/access_import/journal_transform.rb]: Not sure if Rails will overwrite these, or how important it is to keep the original dates

File: lib/access_import/journal_transform.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    }
+  end
+
+  CATEGORY_MAP = {
+    1 => :telephone,
+    2 => :conversation,
+    3 => :email,
+    4 => :feedback,
+    5 => :file
+  }.freeze
```
Comment (toupeira) [inline lib/access_import/journal_transform.rb]: I would always put constants at the top of the class, or at least before method definitions.

File: lib/tasks/setup.rake [hunk 8]
```diff
@@ -7,4 +7,13 @@ namespace :setup do
       warn 'No email set, run `rails setup:superadmin email=email@test.com`'
     end
   end
+
+  desc 'Create a superadmin with factory'
+  task superadmin_factory: :environment do
+    if ENV['email'].present?
+      FactoryGirl.create(:user, email: ENV['email'], password: 'asdfasdf', role: 'superadmin')
+    else
+      warn 'No email set, run `rails setup:superadmin email=email@test.com`'
```
Comment (toupeira) [inline lib/tasks/setup.rake]: `setup:superadmin_factory`

File: app/models/journal.rb [hunk 9]
```diff
@@ -1,10 +1,18 @@
...
   belongs_to :journalable, polymorphic: true, required: false
 
   default_scope { order(created_at: :desc) }
 
   CATEGORIES = [:telephone, :conversation, :email, :feedback, :file].freeze
 
   validates :category, presence: true
+
+  def user
+    super || User.deleted.find(user_id)
```
Comment (toupeira) [inline app/models/journal.rb]: Maybe use `find_by(id: user_id)` to avoid an exception if the user doesn't exist anymore (to get the same behaviour as `super`)

File: app/models/journal.rb [hunk 10]
```diff
@@ -1,10 +1,18 @@
 class Journal < ApplicationRecord
+  include ImportRelation
+
   belongs_to :user
+  belongs_to :assignment, required: false
```
Comment (toupeira) [inline app/models/journal.rb]: From the docs:

> NOTE: `required` is set to `true` by default and is deprecated. If you don't want to have association presence validated, use `optional: true`.

File: app/views/journals/index.html.slim [hunk 11]
```diff
@@ -16,11 +16,11 @@ h1= t_title(:index, Journal)
...
               td= record.subject
-              td= record.body.truncate(40) if record.body
+              td= record.body&.truncate(40)
               td= link_to record.user.full_name, record.user.profile
-              td= link_to t_action(:show), client_journal_path(@journaled, record)
-              td= link_to t_action(:edit), edit_client_journal_path(@journaled, record)
-              td= link_to t_action(:destroy), client_journal_path(@journaled, record), confirm_deleting(record)
+              td= link_to t_action(:show), params.to_unsafe_hash.merge(action: :show, id: record.id)
+              td= link_to t_action(:edit), params.to_unsafe_hash.merge(action: :edit, id: record.id)
+              td= link_to t_action(:destroy), params.to_unsafe_hash.merge(action: :destroy, id: record.id), confirm_deleting(record)
```
Comment (toupeira) [inline app/views/journals/index.html.slim]: Can you try using `url_for(action: :show, id: record.id)` etc. here?

Or actually, I think you can even do `link_to ..., action: :show, id: record.id` without an explicit `url_for` call, though you need to wrap the params in braces for destroy since you're also passing HTML options there.

**Comment by toupeira** [comment]
The whole import code is very complex and could use some tests, but I'm not sure how worthwhile that is since it probably won't be easy :)

File: lib/access_import/journal_transform.rb (hunk 3 above)
Comment (casaper) [reply lib/access_import/journal_transform.rb]: Right. Ill make it by inheritance. It's the same in all those parameter preparers

File: app/views/journals/index.html.slim (hunk 11 above)
Comment (casaper) [reply app/views/journals/index.html.slim]: The think is, that the partial is used from two different controllers (clients and volunteers), so the partial needs to make two different urls.
I find the generally phantastic Rails documentation quite a bit improvable if it comes to the link_to helper. There is a quadzillion different ways to use the thing.

**Comment by casaper** [comment]
> The whole import code is very complex and could use some tests, but I'm not sure how worthwhile that is since it probably won't be easy :)

I'm aware of this, and I completeley agree.
To test this, I have one Idea in mind, but It would definiteley it's own story, because it'll be complex as well:

The legacy MS Access application has exports that spit out MS Excel files containing eg. All Volunteers, All Active Volunteers, same for Clients and so on. Those could be converted to CSV and then tests could randomly pick rows from them and check if the imported counterpart can be found in our application.

File: README.md [hunk 12]
```diff
@@ -29,15 +39,39 @@ Use `rails db:seed` to get these users:
   - email: department_manager@example.com
   - password: asdfasdf
 
-## Sort locale yaml files
+### Create initial superadmin account
+
+Sends a invitaion email to the email address, so that the account can be activated:
```
Comment (toupeira) [inline README.md]: "a**n** invita**t**ion" 👮

File: app/controllers/journals_controller.rb [hunk 13]
```diff
@@ -37,7 +37,7 @@ def update
 
   def destroy
     @journal.destroy
-    redirect_to @journaled, make_notice
+    redirect_to request.referer, make_notice
```
Comment (toupeira) [inline app/controllers/journals_controller.rb]: `redirect_to :back` does the same, with this caveat:

> When using redirect_to :back, if there is no referrer, ActionController::RedirectBackError will be raised. You may specify some fallback behavior for this case by rescuing ActionController::RedirectBackError.

Referer might be empty for various reasons (e.g. privacy extensions), it would probably be sufficient to catch `ActionController::RedirectBackError` in `ApplicationController` and redirect to the root as a fallback.

File: app/models/concerns/deleted_user_relation_fallback.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+module DeletedUserRelationFallback
+  extend ActiveSupport::Concern
+
+  included do
+    def user
+      super || User.deleted.find_by(id: user_id)
+    end
+  end
+end
```
Comment (toupeira) [inline app/models/concerns/deleted_user_relation_fallback.rb]: Had a look at the paranoia README to see if there's a more general way to deal with this, one option would be to specify a scope on the association:

> If you want to include associated soft-deleted objects, you can (un)scope the association:
> 
> ``` ruby
> class Person < ActiveRecord::Base
>   belongs_to :group, -> { with_deleted }
> end
> 
> Person.includes(:group).all
> ```

Should be more reliable, and also avoids running another DB query for deleted users.

File: app/views/journals/_journal_block.html.slim [hunk 15]
```diff
@@ -6,7 +6,7 @@
 #journalBlock.collapse
   .row
     .col-xs-12
-      = button_link t('add_journal_entry'), { controller: 'journals', action: 'new', "#{controller_name.singularize}_id": journaled_for.id }, 'success'
+      = button_link t('add_journal_entry'), url_for([journaled_for, Journal]) + '/new', 'success'
```
Comment (toupeira) [inline app/views/journals/_journal_block.html.slim]: I think `url_for([journaled_for, Journal], action: :new)` should work here as well.

File: lib/tasks/setup.rake [hunk 16]
```diff
@@ -8,12 +8,12 @@ namespace :setup do
     end
   end
 
-  desc 'Create a superadmin with factory'
-  task superadmin_factory: :environment do
+  desc 'Create Superadmin account ready to use (no activation needed)'
+  task superadmin_initialized: :environment do
     if ENV['email'].present?
       FactoryGirl.create(:user, email: ENV['email'], password: 'asdfasdf', role: 'superadmin')
```
Comment (toupeira) [inline lib/tasks/setup.rake]: I think it would be better to just use `User.create` here, FactoryGirl isn't loaded on production and it seems the factory doesn't actually do anything more here (it only sets these three fields as well).

File: lib/tasks/access.rake [hunk 17]
```diff
@@ -1,11 +1,15 @@
...
-    @acimport.make_journal
+    if ENV['file'].present?
+      @acimport = AccessImport.new(ENV['file'])
+      @acimport.make_departments
+      @acimport.make_clients
+      @acimport.make_volunteers
+      @acimport.make_assignments
+      @acimport.make_journal
+    else
+      puts 'No access file set. run "rails access:import file=path/to/access_file.accdb"'
```
Comment (toupeira) [inline lib/tasks/access.rake]: Use `warn`?

File: app/controllers/journals_controller.rb (hunk 13 above)
Comment (casaper) [reply app/controllers/journals_controller.rb]: I used that first, and had exactly that error.

File: app/views/journals/_journal_block.html.slim (hunk 15 above)
Comment (casaper) [reply app/views/journals/_journal_block.html.slim]: I tryed that and for some reason I got a `wrong number of arguments (given 2, expected 0..1)` on that. Thats why I ended up like this.
